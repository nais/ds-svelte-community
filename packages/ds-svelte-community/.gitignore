.DS_Store
node_modules
/build
/dist
/.svelte-kit
/package
.env
.env.*
!.env.example
vite.config.js.timestamp-*
vite.config.ts.timestamp-*
/storybook-static
/src/testlib/*.d.ts
!/src/testlib/ambient.d.ts
