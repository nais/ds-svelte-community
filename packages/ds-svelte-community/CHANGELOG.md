# @nais/ds-svelte-community

## 1.0.0-next.1

### Patch Changes

- 77a004a: Bump dependencies

## 1.0.0-next.0

### Major Changes

- edcf0c8: Support Svelte 5 and make it major
