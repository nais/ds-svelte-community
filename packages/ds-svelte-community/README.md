# ds-svelte-community

## Developing this package

Once you've created a project and installed dependencies with `bun install`, start a development server:

> [!TIP]
> You might need to build the other packages first. Run `bun run build:libs` in the root of the project.

```bash
bun run dev # or
bun run test
```

Everything inside `src/lib` is part of the library, everything inside `src/routes` can be used as a showcase or preview app.

## Adding new components

In general, we try to create components that behave like the components in the [React version](https://github.com/navikt/aksel/tree/main/%40navikt/core/react), with a few exceptions:

- Default user facing texts are in English, not Norwegian.
- We try to make the components feel more like Svelte components, and less like React components.

We strive to generate HTML equal to the HTML generated by the React version of the component library.

We ensure this by adding a test for each component, that renders the component and compares the HTML to the HTML generated by the React version.

```typescript
import { bunmatch } from "$testlib/bunmatch";
import { Loader as ReactLoader } from "@navikt/ds-react";
import { cleanup, render } from "@testing-library/svelte/svelte5";
import { afterEach, describe, expect, it } from "bun:test";
import Loader from "./Loader.svelte";
import type { Props } from "./type";

describe("Loader", () => {
	it("renders with HTML similar to ds-react", async () => {
		// Often we can define one structure with props, used by both components.
		// If there's differences, we should pass separate props to each component.
		const props: Props = {
			title: "Loading",
		};
		expect(
			// bunmatch is a custom matcher that returns true if they are similar.
			// bunmatch will do some alterations to the HTML, e.g. sort attributes, classes and format the html.
			// The API is a bit of a hack until Bun supports custom matchers.
			// The first argument is the Svelte component, the second is the React component.
			// The third argument is an object with props and options.
			await bunmatch(render(Loader, props), ReactLoader, {
				props: {
					// Spread the props object, so we can add more props if needed.
					...props,
				},
				opts: {
					// There's a few functions that alters the HTML. E.g. this one removes any attribute where the function returns false.
					compareAttrs(node, attr) {
						const tagName = node.tagName.toLowerCase();
						// Title ids are unique
						if (tagName == "svg" && ["aria-labelledby"].includes(attr)) {
							return false;
						}

						if (tagName == "title" && attr == "id") {
							return false;
						}
						return true;
					},
				},
			}),
		).toBeTrue(); // bunmatch returns true if the HTML is similar.
	});

	afterEach(cleanup);
});
```

## Building

To build the library:

```bash
bun run package
```

To create a production version of the showcase app:

```bash
bun run build
```

### TODO

- Stories doesn't include js/ts used in e.g. {#each} blocks;
