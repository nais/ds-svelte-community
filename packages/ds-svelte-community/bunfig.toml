telemetry = false

[test]
preload = ["./src/testlib/bun_setup.ts", "./src/testlib/test_helper.ts"]
root = "./src/"
