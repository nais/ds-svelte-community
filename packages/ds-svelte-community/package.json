{
	"name": "@nais/ds-svelte-community",
	"version": "1.0.0-next.31",
	"homepage": "https://nais.github.io/ds-svelte-community",
	"repository": {
		"type": "git",
		"url": "git+https://github.com/nais/ds-svelte-community.git",
		"directory": "packages/ds-svelte-community"
	},
	"license": "MIT",
	"scripts": {
		"dev": "bunx --bun vite dev",
		"build": "bunx --bun vite build && bun run package",
		"build:package": "bun run package",
		"build:docs": "bunx --bun vite build",
		"preview": "bunx --bun vite preview",
		"package": "svelte-package && publint",
		"prepublishOnly": "bun run test:unit && bun run check && bun run lint && bun run package",
		"check": "svelte-check --tsconfig ./tsconfig.json",
		"check:watch": "svelte-check --tsconfig ./tsconfig.json --watch",
		"lint": "prettier --check . && eslint .",
		"format": "prettier --write .",
		"generate-icons": "./hack/generate-icons.sh",
		"prepare": "svelte-kit sync || echo ''",
		"test": "bun run test:unit",
		"test:unit": "bun test --conditions browser",
		"test:unit:watch": "bun test --conditions browser --watch",
		"sync": "svelte-kit sync"
	},
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"import": "./dist/index.js",
			"svelte": "./dist/index.js"
		},
		"./icons": {
			"types": "./dist/icons/index.d.ts",
			"import": "./dist/icons/index.js",
			"svelte": "./dist/icons/index.js"
		},
		"./experimental": {
			"types": "./dist/experimental/index.d.ts",
			"import": "./dist/experimental/index.js",
			"svelte": "./dist/experimental/index.js"
		},
		"./css": {
			"default": "./dist/css/index.css"
		},
		"./*": {
			"types": "./dist/*",
			"import": "./dist/*",
			"svelte": "./dist/*"
		}
	},
	"files": [
		"dist",
		"!dist/**/*.test.*",
		"!dist/**/*.spec.*"
	],
	"peerDependencies": {
		"vite": "^5.4.9 || ^6.0.3",
		"svelte": "^5.20.0"
	},
	"optionalDependencies": {
		"svelte-floating-ui": "^1.6.2"
	},
	"devDependencies": {
		"@nais/ds-svelte-community-preprocess-svelte": "workspace:*",
		"@nais/vite-plugin-svelte-docs": "workspace:*",
		"@navikt/aksel-icons": "^7.19.0",
		"@navikt/ds-react": "^7.19.0",
		"@navikt/ds-tailwind": "^7.19.0",
		"@sveltejs/adapter-static": "^3.0.8",
		"@sveltejs/kit": "^2.20.5",
		"@sveltejs/package": "^2.3.10",
		"@sveltejs/vite-plugin-svelte": "^5.0.3",
		"@tailwindcss/vite": "^4.1.3",
		"@testing-library/svelte": "^5.2.7",
		"@types/bun": "^1.2.9",
		"@types/diff": "^7.0.2",
		"@types/react-dom": "^19.1.2",
		"autoprefixer": "^10.4.21",
		"diff": "^7.0.0",
		"eslint": "^9.24.0",
		"eslint-config-prettier": "^10.1.1",
		"eslint-plugin-svelte": "^3.5.1",
		"global-jsdom": "^26.0.0",
		"jsdom": "^26.0.0",
		"marked": "^15.0.8",
		"mdsvex": "^0.12.3",
		"prettier": "^3.5.3",
		"prettier-plugin-svelte": "^3.3.3",
		"publint": "^0.3.11",
		"react": "^19.1.0",
		"react-dom": "^19.1.0",
		"rehype-autolink-headings": "^7.1.0",
		"rehype-slug": "^6.0.0",
		"resize-observer-polyfill": "^1.5.1",
		"shiki": "^3.2.2",
		"svelte": "5.25.10",
		"svelte-check": "^4.1.5",
		"svelte-floating-ui": "^1.6.2",
		"svgo": "3.3.2",
		"tailwindcss": "^4.1.3",
		"typescript": "5.8.3",
		"typescript-eslint": "^8.29.1",
		"vite": "^6.2.6"
	},
	"dependencies": {
		"@navikt/ds-css": "^7.19.0",
		"@navikt/ds-tokens": "^7.19.0",
		"clsx": "^2.1.1"
	},
	"svelte": "./dist/index.js",
	"types": "./dist/index.d.ts",
	"type": "module",
	"overrides": {
		"svelte": "$svelte"
	}
}
