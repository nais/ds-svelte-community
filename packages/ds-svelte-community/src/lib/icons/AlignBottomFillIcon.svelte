<script lang="ts" context="module">
	import newUniqueId from "$lib/components/local-unique-id";
	import type { SVGAttributes } from "svelte/elements";
</script>

<script lang="ts">
	/**
	 * Title of the icon, used for accessibility
	 */
	export let title = "";

	type $$Props = SVGAttributes<SVGElement> & {
		title?: string;
	};

	const id = newUniqueId();
</script>

<svg
	xmlns="http://www.w3.org/2000/svg"
	width="1em"
	height="1em"
	fill="none"
	focusable="false"
	role="img"
	viewBox="0 0 24 24"
	aria-labelledby={title ? id : undefined}
	{...$$restProps}
>
	{#if title}
		<title {id}>{title}</title>
	{/if}
	<slot />
	<path
		fill="currentColor"
		fill-rule="evenodd"
		d="M6 17.75a.75.75 0 0 1-.75-.75V4A.75.75 0 0 1 6 3.25h4a.75.75 0 0 1 .75.75v13a.75.75 0 0 1-.75.75H6ZM3.25 20a.75.75 0 0 1 .75-.75h16a.75.75 0 0 1 0 1.5H4a.75.75 0 0 1-.75-.75Zm10-3c0 .414.336.75.75.75h4a.75.75 0 0 0 .75-.75v-7a.75.75 0 0 0-.75-.75h-4a.75.75 0 0 0-.75.75v7Z"
		clip-rule="evenodd"
	/>
</svg>
