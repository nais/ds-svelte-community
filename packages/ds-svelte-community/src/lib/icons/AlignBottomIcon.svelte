<script lang="ts" context="module">
	import newUniqueId from "$lib/components/local-unique-id";
</script>

<script lang="ts">
	import type { Snippet } from "svelte";

	type Props = {
		/**
		 * Title of the icon, used for accessibility
		 */
		title?: string;

		children?: Snippet;

		[key: string]: unknown;
	};

	let { title, children, ...restProps }: Props = $props();

	const id = newUniqueId();
</script>

<svg
	xmlns="http://www.w3.org/2000/svg"
	width="1em"
	height="1em"
	fill="none"
	focusable="false"
	role="img"
	viewBox="0 0 24 24"
	aria-labelledby={title ? id : undefined}
	{...restProps}
>
	{#if title}
		<title {id}>{title}</title>
	{/if}
	{#if children}
		{@render children()}
	{/if}
	<path
		fill="currentColor"
		fill-rule="evenodd"
		d="M6 17.75a.75.75 0 0 1-.75-.75V4A.75.75 0 0 1 6 3.25h4a.75.75 0 0 1 .75.75v13a.75.75 0 0 1-.75.75H6Zm.75-1.5h2.5V4.75h-2.5v11.5ZM3.25 20a.75.75 0 0 1 .75-.75h16a.75.75 0 0 1 0 1.5H4a.75.75 0 0 1-.75-.75Zm10-3c0 .414.336.75.75.75h4a.75.75 0 0 0 .75-.75v-7a.75.75 0 0 0-.75-.75h-4a.75.75 0 0 0-.75.75v7Zm4-.75h-2.5v-5.5h2.5v5.5Z"
		clip-rule="evenodd"
	/>
</svg>
