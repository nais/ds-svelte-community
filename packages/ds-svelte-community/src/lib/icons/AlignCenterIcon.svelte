<script lang="ts" context="module">
	import newUniqueId from "$lib/components/local-unique-id";
	import type { SVGAttributes } from "svelte/elements";
</script>

<script lang="ts">
	/**
	 * Title of the icon, used for accessibility
	 */
	export let title = "";

	type $$Props = SVGAttributes<SVGElement> & {
		title?: string;
	};

	const id = newUniqueId();
</script>

<svg
	xmlns="http://www.w3.org/2000/svg"
	width="1em"
	height="1em"
	fill="none"
	focusable="false"
	role="img"
	viewBox="0 0 24 24"
	aria-labelledby={title ? id : undefined}
	{...$$restProps}
>
	{#if title}
		<title {id}>{title}</title>
	{/if}
	<slot />
	<path
		fill="currentColor"
		fill-rule="evenodd"
		d="M12 3.25a.75.75 0 0 1 .75.75v1.25h5.75a.75.75 0 0 1 .75.75v4a.75.75 0 0 1-.75.75h-5.75v2.5h2.75a.75.75 0 0 1 .75.75v4a.75.75 0 0 1-.75.75h-2.75V20a.75.75 0 0 1-1.5 0v-1.25H8.5a.75.75 0 0 1-.75-.75v-4a.75.75 0 0 1 .75-.75h2.75v-2.5H5.5a.75.75 0 0 1-.75-.75V6a.75.75 0 0 1 .75-.75h5.75V4a.75.75 0 0 1 .75-.75Zm-.75 11.5h-2v2.5h2v-2.5Zm1.5 2.5h2v-2.5h-2v2.5Zm-6.5-10.5h5v2.5h-5v-2.5Zm6.5 0v2.5h5v-2.5h-5Z"
		clip-rule="evenodd"
	/>
</svg>
