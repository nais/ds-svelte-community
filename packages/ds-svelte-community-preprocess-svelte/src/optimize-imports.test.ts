import { describe, expect, it } from "bun:test";
import { optimizeImports } from ".";

describe("optimize-imports", () => {
	it("replaces components", () => {
		const script = `
		import { Button, Tab, Alert, Detail } from "@nais/ds-svelte-community";
		`;

		const pp = optimizeImports();

		expect(
			pp.script!({ content: script, filename: "src/+page.svelte" } as never),
		).toMatchSnapshot();
	});

	it("replaces components with options", () => {
		const script = `
		import { Button, Tab, Alert, Detail } from "$lib";
		`;

		const pp = optimizeImports({
			componentsImportPrefix: {
				match: "$lib",
				prefix: "$lib/",
			},
		});

		expect(
			pp.script!({ content: script, filename: "src/+page.svelte" } as never),
		).toMatchSnapshot();
	});

	it("replaces icons", () => {
		const script = `
		import { AirplaneIcon, CarIcon } from "@nais/ds-svelte-community/icons";
		`;

		const pp = optimizeImports();

		expect(
			pp.script!({ content: script, filename: "src/+page.svelte" } as never),
		).toMatchSnapshot();
	});

	it("replaces icons with options", () => {
		const script = `
		import { AirplaneIcon, CarIcon } from "$lib/icons";
		`;

		const pp = optimizeImports({
			iconImportPrefix: {
				match: "$lib/icons",
				prefix: "$lib/",
			},
		});

		expect(
			pp.script!({ content: script, filename: "src/+page.svelte" } as never),
		).toMatchSnapshot();
	});

	it("replaces mixed", () => {
		const script = `
		import { Button, Tab, Alert, Detail } from "@nais/ds-svelte-community";
		import { AirplaneIcon, CarIcon } from "@nais/ds-svelte-community/icons";
		`;

		const pp = optimizeImports();

		expect(
			pp.script!({ content: script, filename: "src/+page.svelte" } as never),
		).toMatchSnapshot();
	});

	it("replaces mixed with options", () => {
		const script = `
		import { Button, Tab, Alert, Detail } from "$lib";
		import { AirplaneIcon, CarIcon } from "$lib/icons";
		`;

		const pp = optimizeImports({
			componentsImportPrefix: {
				match: "$lib",
				prefix: "$lib/",
			},
			iconImportPrefix: {
				match: "$lib/icons",
				prefix: "$lib/",
			},
		});

		expect(
			pp.script!({ content: script, filename: "src/+page.svelte" } as never),
		).toMatchSnapshot();
	});

	it("doesn't do anything when importing star", () => {
		const script = `
		import * as icons from "$lib/icons";
		`;

		const pp = optimizeImports({
			componentsImportPrefix: {
				match: "$lib",
				prefix: "$lib/",
			},
			iconImportPrefix: {
				match: "$lib/icons",
				prefix: "$lib/",
			},
		});

		const d = pp.script!({ content: script, filename: "src/+page.svelte" } as never);
		console.log(d);
		expect(d).toMatchSnapshot();
	});
});
