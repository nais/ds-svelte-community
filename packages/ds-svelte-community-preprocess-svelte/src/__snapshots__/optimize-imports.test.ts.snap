// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`optimize-imports replaces components 1`] = `
{
  "code": 
"
		import Button from "@nais/ds-svelte-community/components/Button/Button.svelte";
import Tab from "@nais/ds-svelte-community/components/Tabs/Tab.svelte";
import Alert from "@nais/ds-svelte-community/components/Alert/Alert.svelte";
import Detail from "@nais/ds-svelte-community/components/typography/Detail/Detail.svelte";
		"
,
}
`;

exports[`optimize-imports replaces components with options 1`] = `
{
  "code": 
"
		import Button from "$lib/components/Button/Button.svelte";
import Tab from "$lib/components/Tabs/Tab.svelte";
import Alert from "$lib/components/Alert/Alert.svelte";
import Detail from "$lib/components/typography/Detail/Detail.svelte";
		"
,
}
`;

exports[`optimize-imports replaces icons 1`] = `
{
  "code": 
"
		import AirplaneIcon from "@nais/ds-svelte-community/icons/AirplaneIcon.svelte";
import CarIcon from "@nais/ds-svelte-community/icons/CarIcon.svelte";
		"
,
}
`;

exports[`optimize-imports replaces icons with options 1`] = `
{
  "code": 
"
		import AirplaneIcon from "$lib/icons/AirplaneIcon.svelte";
import CarIcon from "$lib/icons/CarIcon.svelte";
		"
,
}
`;

exports[`optimize-imports replaces mixed 1`] = `
{
  "code": 
"
		import Button from "@nais/ds-svelte-community/components/Button/Button.svelte";
import Tab from "@nais/ds-svelte-community/components/Tabs/Tab.svelte";
import Alert from "@nais/ds-svelte-community/components/Alert/Alert.svelte";
import Detail from "@nais/ds-svelte-community/components/typography/Detail/Detail.svelte";
		import AirplaneIcon from "@nais/ds-svelte-community/icons/AirplaneIcon.svelte";
import CarIcon from "@nais/ds-svelte-community/icons/CarIcon.svelte";
		"
,
}
`;

exports[`optimize-imports replaces mixed with options 1`] = `
{
  "code": 
"
		import Button from "$lib/components/Button/Button.svelte";
import Tab from "$lib/components/Tabs/Tab.svelte";
import Alert from "$lib/components/Alert/Alert.svelte";
import Detail from "$lib/components/typography/Detail/Detail.svelte";
		import AirplaneIcon from "$lib/icons/AirplaneIcon.svelte";
import CarIcon from "$lib/icons/CarIcon.svelte";
		"
,
}
`;

exports[`optimize-imports doesn't do anything when importing star 1`] = `
{
  "code": 
"
		import * as icons from "$lib/icons";
		"
,
}
`;
